velox/exec/WindowPartition.cpp

```diff
@@ -29,6 +29,51 @@ WindowPartition::WindowPartition(
   for (int i = 0; i < inputMapping_.size(); i++) {
     columns_.emplace_back(data_->columnAt(inputMapping_[i]));
   }
+
+  complete_ = true;
+}
+
+WindowPartition::WindowPartition(
+    RowContainer* data,
+    const std::vector<column_index_t>& inputMapping,
+    const std::vector<std::pair<column_index_t, core::SortOrder>>& sortKeyInfo)
+    : data_(data), inputMapping_(inputMapping), sortKeyInfo_(sortKeyInfo) {
+  for (int i = 0; i < inputMapping_.size(); i++) {
+    columns_.emplace_back(data_->columnAt(inputMapping_[i]));
+  }
+
+  rows_.clear();
+  partition_ = folly::Range(rows_.data(), rows_.size());
+  complete_ = false;
+  partial_ = true;
+}
+
+void WindowPartition::addRows(const std::vector<char*>& rows) {
+  VELOX_CHECK(partial_, "Current WindowPartition should be partial.");
+  rows_.insert(rows_.end(), rows.begin(), rows.end());
+  partition_ = folly::Range(rows_.data(), rows_.size());
+}
+
+void WindowPartition::eraseRows(vector_size_t numRows) {
+  data_->eraseRows(folly::Range<char**>(rows_.data(), numRows));
+  rows_.erase(rows_.begin(), rows_.begin() + numRows);
+  partition_ = folly::Range(rows_.data(), rows_.size());
+}
+
+void WindowPartition::removeProcessedRows(vector_size_t numRows) {
+  VELOX_CHECK(partial_, "Current WindowPartition should be partial.");
+  if (!complete_ || (complete_ && rows_.size() >= numRows)) {
```

Comment:
Why we don't erase on a complete partition? We only set a partition to complete after we detect a new partition?